vllm/model_executor/models/mlp_speculator.py

```diff
@@ -0,0 +1,151 @@
+        self.inner_dim = config.inner_dim if config.inner_dim != 0 \
+            else config.emb_dim
+        self.emb = nn.ModuleList([
+            VocabParallelEmbedding(config.vocab_size,
+                                   self.inner_dim,
+                                   org_num_embeddings=config.vocab_size)
+            for _ in range(config.n_predict)
+        ])
+
+        self.proj = nn.ModuleList([
+            nn.Linear((self.emb_dim if i == 0 else self.inner_dim),
+                      self.inner_dim,
+                      bias=False) for i in range(config.n_predict)
+        ])
+
+        self.head = nn.ModuleList([
+            nn.Linear(self.inner_dim, self.vocab_size, bias=False)
+            for _ in range(config.n_predict)
+        ])
+        self.ln = nn.ModuleList([
+            MLPSpeculatorLayerNorm(self.inner_dim)
+            for _ in range(config.n_predict)
+        ])
+
+        self.state_weight = 0.5**(0.5 / config.n_predict)
+        self.emb_weight = math.sqrt(
+            (1 - self.state_weight**2) * (self.inner_dim / 2))
+        self.activation = nn.GELU()
+        self.config = config
+        self.logits_processor = LogitsProcessor(config.vocab_size,
+                                                config.vocab_size, 1.0)
+        self.sampler = Sampler()
+
+    def generate_proposals(
+        self,
+        input_ids: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> torch.Tensor:
+
+        if self.first_decode_step:
+            self.first_decode_step = False
+        else:
+            self.previous_hidden_state = self.previous_hidden_state.reshape(
+                -1, self.n_predict + 1, self.previous_hidden_state.size(1))
+            self.previous_hidden_state = self.previous_hidden_state.gather(
+                1, (self.accepted_token_lengths - 1)[:, None, None].expand(
+                    -1, 1,
+                    self.previous_hidden_state.size(2))).squeeze(1)  # b x d
+
+        # b x 1 x d
+        self.previous_hidden_state = self.previous_hidden_state.reshape(
+            self.previous_hidden_state.size(0), 1,
+            self.previous_hidden_state.size(1))
+
+        # b x 1
+        last_tokens = input_ids.reshape(-1, 1)
+
+        next_tokens = []
+
+        for head_index in range(self.n_predict):
+
+            # Project and predict
+            z = self.emb[head_index](last_tokens)  # b k d
+            state = self.proj[head_index](self.previous_hidden_state)
+
+            # Weighted add of state_weight*state and emb_weight*z
+            # Let subsequent LN take care of denominator
+            # state_weight is close to 1, so shouldn't be any precision issues
+            state = torch.add(state,
+                              z,
+                              alpha=self.emb_weight / self.state_weight)
+
+            state = self.activation(self.ln[head_index](state))  # b k d
+            # todo: not yet supporting top_k_tokens_per_head
```

Comment:
Without this, we are assuming the set of `top_k_tokens_per_head` are each 1 for every head. We may want to see how this affects performance (acceptance rate) as typically we create a tree and traverse the tree to find the optimal candidate.

Reply:
So, `n_candidates` is also not used? (it is assumed to be equal 1?)